File: app/src/main/java/com/example/smishingdetectionapp/SmsReceiver.java
//package com.example.smishingdetectionapp;
//import android.Manifest;
//
//import android.app.PendingIntent;
//import android.content.BroadcastReceiver;
//import android.content.Context;
//import android.content.Intent;
//import android.content.pm.PackageManager;
//import android.os.Build;
//import android.telephony.SmsMessage;
//import android.util.Log;
//
//import androidx.core.app.NotificationCompat;
//import androidx.core.app.NotificationManagerCompat;
//import androidx.core.content.ContextCompat;
//
//public class SmsReceiver extends BroadcastReceiver {
//    private static final String TAG = "SmsReceiver";
//    private static final String CHANNEL_ID = "smishing_alert_channel";
//
//    @Override
//    public void onReceive(Context context, Intent intent) {
//        if (intent.getAction().equals("android.provider.Telephony.SMS_RECEIVED")) {
//            if (ContextCompat.checkSelfPermission(context, Manifest.permission.RECEIVE_SMS) != PackageManager.PERMISSION_GRANTED ||
//                    ContextCompat.checkSelfPermission(context, Manifest.permission.READ_SMS) != PackageManager.PERMISSION_GRANTED) {
//                Log.e(TAG, "SMS permissions are not granted");
//                return;
//            }
//
//            SmsMessage[] messages = extractSmsMessages(intent);
//            if (messages != null) {
//                for (SmsMessage sms : messages) {
//                    if (sms != null) {
//                        String sender = sms.getDisplayOriginatingAddress();
//                        String messageBody = sms.getMessageBody();
//
//                        if (sender != null && messageBody != null) {
//                            boolean isSuspicious = SmishingDetector.isSmishingMessage(messageBody.toLowerCase());
//                            if (isSuspicious) {
//                                showNotification(context, sender, messageBody);
//                            }
//                        } else {
//                            Log.e(TAG, "Sender or message body is null");
//                        }
//                    } else {
//                        Log.e(TAG, "SmsMessage is null");
//                    }
//                }
//            } else {
//                Log.e(TAG, "No messages found");
//            }
//        }
//    }
//
//    private SmsMessage[] extractSmsMessages(Intent intent) {
//        Object[] pdus = (Object[]) intent.getExtras().get("pdus");
//        if (pdus == null) {
//            Log.e(TAG, "No PDUs found in the intent");
//            return null;
//        }
//
//        SmsMessage[] messages = new SmsMessage[pdus.length];
//        String format = intent.getStringExtra("format");
//        if (format == null) {
//            format = "3gpp"; // Default format
//        }
//
//        for (int i = 0; i < pdus.length; i++) {
//            try {
//                messages[i] = SmsMessage.createFromPdu((byte[]) pdus[i], format);
//            } catch (Exception e) {
//                Log.e(TAG, "Error creating SmsMessage from PDU", e);
//            }
//        }
//
//        return messages;
//    }
//
//    private void showNotification(Context context, String sender, String messageBody) {
//        Intent intent = new Intent(context, SMSMessageDetailActivity.class);
//        intent.putExtra("SMS_SENDER", sender);
//        intent.putExtra("SMS_BODY", messageBody);
//        PendingIntent pendingIntent = PendingIntent.getActivity(context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE);
//
//        NotificationCompat.Builder builder = new NotificationCompat.Builder(context, CHANNEL_ID)
//                .setSmallIcon(R.drawable.hardhat_logo)
//                .setContentTitle("Suspicious Message Detected")
//                .setContentText("Possible phishing message from " + sender)
//                .setPriority(NotificationCompat.PRIORITY_HIGH)
//                .setContentIntent(pendingIntent)
//                .setAutoCancel(true);
//
//        NotificationManagerCompat notificationManager = NotificationManagerCompat.from(context);
//        if (ContextCompat.checkSelfPermission(context, Manifest.permission.POST_NOTIFICATIONS) == PackageManager.PERMISSION_GRANTED) {
//            notificationManager.notify((int) System.currentTimeMillis(), builder.build());
//        }
//    }
//}

package com.example.smishingdetectionapp;

import android.Manifest;
import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.telephony.SmsMessage;
import android.util.Log;

import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationManagerCompat;
import androidx.core.content.ContextCompat;

public class SmsReceiver extends BroadcastReceiver {
    private static final String TAG = "SmsReceiver";
    private static final String CHANNEL_ID = "smishing_alert_channel";

    @Override
    public void onReceive(Context context, Intent intent) {
        if (intent.getAction().equals("android.provider.Telephony.SMS_RECEIVED")) {
            // Check SMS permissions
            if (ContextCompat.checkSelfPermission(context, Manifest.permission.RECEIVE_SMS) != PackageManager.PERMISSION_GRANTED ||
                    ContextCompat.checkSelfPermission(context, Manifest.permission.READ_SMS) != PackageManager.PERMISSION_GRANTED) {
                Log.e(TAG, "SMS permissions are not granted");
                return;
            }

            SmsMessage[] messages = extractSmsMessages(intent);
            if (messages != null) {
                for (SmsMessage sms : messages) {
                    if (sms != null) {
                        String sender = sms.getDisplayOriginatingAddress();
                        String messageBody = sms.getMessageBody();

                        if (sender != null && messageBody != null) {
                            boolean isSuspicious = SmishingDetector.isSmishingMessage(messageBody.toLowerCase());
                            if (isSuspicious) {
                                showNotification(context, sender, messageBody);
                            }
                        } else {
                            Log.e(TAG, "Sender or message body is null");
                        }
                    } else {
                        Log.e(TAG, "SmsMessage is null");
                    }
                }
            } else {
                Log.e(TAG, "No messages found");
            }
        }
    }

    private SmsMessage[] extractSmsMessages(Intent intent) {
        Object[] pdus = (Object[]) intent.getExtras().get("pdus");
        if (pdus == null) {
            Log.e(TAG, "No PDUs found in the intent");
            return null;
        }

        SmsMessage[] messages = new SmsMessage[pdus.length];
        String format = intent.getStringExtra("format");
        for (int i = 0; i < pdus.length; i++) {
            messages[i] = SmsMessage.createFromPdu((byte[]) pdus[i], format);
        }
        return messages;
    }

    private void showNotification(Context context, String sender, String messageBody) {
        Intent intent = new Intent(context, SMSMessageDetailActivity.class);
        intent.putExtra("SMS_SENDER", sender);
        intent.putExtra("SMS_BODY", messageBody);
        PendingIntent pendingIntent = PendingIntent.getActivity(context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE);

        NotificationCompat.Builder builder = new NotificationCompat.Builder(context, CHANNEL_ID)
                .setSmallIcon(R.drawable.hardhat_logo)
                .setContentTitle("Suspicious Message Detected")
                .setContentText("Possible phishing message from " + sender)
                .setPriority(NotificationCompat.PRIORITY_HIGH)
                .setContentIntent(pendingIntent)
                .setAutoCancel(true);

        NotificationManagerCompat notificationManager = NotificationManagerCompat.from(context);
        if (ContextCompat.checkSelfPermission(context, Manifest.permission.POST_NOTIFICATIONS) == PackageManager.PERMISSION_GRANTED) {
            notificationManager.notify((int) System.currentTimeMillis(), builder.build());
        }
    }
}

File: app/src/main/java/SmsReceiver.kt
package com.example.smishingdetectionapp

import com.example.smishingdetectionapp.AESUtil
import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.os.Build
import android.telephony.SmsMessage
import android.util.Log
import android.widget.Toast
import androidx.core.app.NotificationCompat
import java.io.IOException
import java.security.GeneralSecurityException

class SmsReceiver : BroadcastReceiver() {

    override fun onReceive(context: Context?, intent: Intent?) {
        Log.d("SmsReceiver", "onReceive called")

        // Ensure context is not null
        if (context == null) {
            Log.e("SmsReceiver", "Context is null, cannot proceed")
            return
        }

        // Check if the received intent is an SMS received action
        if (intent?.action == "android.provider.Telephony.SMS_RECEIVED") {
            val bundle = intent.extras
            if (bundle != null) {
                Log.d("SmsReceiver", "Bundle is not null")
                val pdus = bundle.get("pdus") as Array<*>

                // Loop through all received PDUs (Protocol Data Units)
                for (pdu in pdus) {
                    val format = bundle.getString("format")
                    val smsMessage = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                        SmsMessage.createFromPdu(pdu as ByteArray, format)
                    } else {
                        SmsMessage.createFromPdu(pdu as ByteArray)
                    }

                    val messageBody = smsMessage.messageBody
                    val sender = smsMessage.originatingAddress

                    Log.d("SmsReceiver", "Received SMS: $messageBody from $sender")

                    // Log the original SMS content
                    Log.d("EncryptionUtil", "Original SMS content: $messageBody")

                    if (messageBody.isNullOrEmpty()) {
                        Log.e("SmsReceiver", "Message body is null or empty before encryption.")
                        return
                    }

                    try {
                        // Encrypt the SMS content
                        Log.d("AESUtil", "Starting encryption process")
                        val encryptedMessage = AESUtil.encrypt(messageBody)
                        if (encryptedMessage.isNullOrEmpty()) {
                            Log.e("SmsReceiver", "Failed to encrypt SMS content: encryptedMessage is null or empty.")
                        } else {
                            Log.d("AESUtil", "Encrypted SMS content: $encryptedMessage")
                        }

                        // Decrypt the SMS content when retrieving
                        Log.d("AESUtil", "Starting decryption process")

                        if (!encryptedMessage.isNullOrEmpty()) {
                            val decryptedMessage = AESUtil.decrypt(encryptedMessage)
                            Log.d("AESUtil", "Decrypted SMS content: $decryptedMessage")

                            // Display the decrypted SMS content with a Toast
                            Toast.makeText(context, "Decrypted SMS Content: $decryptedMessage", Toast.LENGTH_SHORT).show()
                        } else {
                            Log.e("SmsReceiver", "Failed to decrypt SMS content: encryptedMessage is null or empty.")
                        }

                    } catch (e: GeneralSecurityException) {
                        Log.e("SmsReceiver", "Error encrypting or decrypting SMS content", e)
                    } catch (e: IOException) {
                        Log.e("SmsReceiver", "Error encrypting or decrypting SMS content", e)
                    }

                    // Check if the SMS message is suspected of smishing
                    if (isSmishingMessage(messageBody)) {
                        Log.d("SmsReceiver", "Smishing detected!")
                        sendNotification(context, sender, messageBody)
                    }
                }
            } else {
                Log.d("SmsReceiver", "Bundle is null")
            }
        }
    }

    // Method to determine if the message is a potential smishing attempt
    private fun isSmishingMessage(message: String): Boolean {
        return message.contains("suspicious link", true) || message.contains("urgent action required", true)
    }

    // Method to send a notification if smishing is detected
    private fun sendNotification(context: Context, sender: String?, message: String) {
        val notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

        // Create a notification channel for Android O and above
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(
                "smishing_channel",
                "Smishing Alerts",
                NotificationManager.IMPORTANCE_HIGH
            )
            notificationManager.createNotificationChannel(channel)
        }

        // Build the notification
        val notification = NotificationCompat.Builder(context, "smishing_channel")
            .setSmallIcon(R.drawable.ic_notification)  // Ensure this resource exists
            .setContentTitle("Smishing Alert!")
            .setContentText("Message from $sender: $message. DO NOT click on the link!")
            .setPriority(NotificationCompat.PRIORITY_HIGH)
            .setAutoCancel(true)
            .build()

        // Display the notification
        notificationManager.notify(1, notification)
    }
}
